src/librustdoc/html/markdown.rs

```diff
@@ -1190,10 +1201,20 @@ crate fn markdown_links(md: &str) -> Vec<(String, Range<usize>)> {
     let iter = Footnotes::new(HeadingLinks::new(p.into_offset_iter(), None, &mut ids));
 
     for ev in iter {
-        if let Event::Start(Tag::Link(_, dest, _)) = ev.0 {
+        if let Event::Start(Tag::Link(kind, dest, _)) = ev.0 {
             debug!("found link: {}", dest);
-            let span = span_for_link(&dest, ev.1);
-            links.push((dest.into_string(), span));
+            let link = dest.into_string();
+            let span = if let LinkType::Shortcut | LinkType::Reference = kind {
+                // NOTE (I'm sorry Joshua):
+                // For reference-style links (where the target is defined in footnote) we want the
+                // span to point to the definition, not the usage in order to generate useful error
+                // messages.
+                let start = link.as_ptr() as usize - md.as_ptr() as usize;
```

Comment:
This actually doesn't work because the parser may clone the string and the pointer will no longer point into `md`. Sad :(

Reply:
Can you open an issue for `pulldown`? I think pointing to the usage is fine for now though.

Reply:
To be honest, I don't consider this a bug in `pulldown`. Current behaviour is strange, but unless `pulldown` can provide two spans for a single event, this is at least consistent enough.

Pointing at the usage feels weird, though. This doesn't point me to what I need to modify:
```
+ LL | /// Links to [a] [link][a]
+    |              ^^^ no item named `ref` in scope
```

But I guess we can live with it if we can't fix it.

On the other hand, if it were possible to look up the link's title in the list of definitions, we could maybe point at the definition itself. I guess we can build a collector pass that build a table from footnotes? I need to look into how pulldows actually parses text and which events are relevant here. That will probably have to wait for next year, though.

Reply:
Hmm, yeah that error is pretty bad. I think I would rather have duplicate errors than an error pointing to the wrong place :/

Reply:
You broke it ;)

Reply:
Oh oof, you're right, this is latest nightly:

```
warning: unresolved link to `std::process::Comman`
 --> link.rs:1:14
  |
1 | //! Links to [a] [link][a]
  |              ^^^ no item named `Comman` in module `process`
  |
  = note: `#[warn(broken_intra_doc_links)]` on by default

warning: unresolved link to `std::process::Comman`
 --> link.rs:1:18
  |
1 | //! Links to [a] [link][a]
  |                  ^^^^^^^^^ no item named `Comman` in module `process`
```

and this is a couple days ago, before the PR removing locate():

```
warning: unresolved link to `std::process::Comman`
 --> link.rs:3:10
  |
3 | //! [a]: std::process::Comman
  |          ^^^^^^^^^^^^^^^^^^^^ no item named `Comman` in module `process`
  |
  = note: `#[warn(broken_intra_doc_links)]` on by default

warning: unresolved link to `std::process::Comman`
 --> link.rs:3:10
  |
3 | //! [a]: std::process::Comman
  |          ^^^^^^^^^^^^^^^^^^^^ no item named `Comman` in module `process`
```